### ISSUE TITLE
When assigning a variable that has already been checked, MyPy insists on redoing the check

### ISSUE THREAD

**Issue by moi15moi** [python/mypy#20481]
**Bug Report**
When assigning a variable that has already been checked, MyPy insists on redoing the check.

**To Reproduce**

See: https://mypy-play.net/?mypy=latest&python=3.12&gist=8fbe0316234691cdc2f8e85e324d56e9
Or
```py
from pathlib import Path


class MyClass:
    def __init__(self, path: Path | None) -> None:
        self.path = path


def test(a: MyClass, b: MyClass):
    if a.path is None:
        raise ValueError(f"a.path cannot be None")
    if b.path is None:
        raise ValueError(f"b.path cannot be None")

    # This line will cause an error for the line "b.path.exists()": error: Item "None" of "Path | None" has no attribute "exists"  [union-attr]
    b = a

    if a.path.exists():
        print("a path exist")

    if b.path.exists():
        print("b path exist")


def main():
    a = MyClass(Path("."))
    b = MyClass(Path("."))

    test(a, b)


if __name__ == "__main__":
    main()
```

**Expected Behavior**

The command ``mypy main.py`` should report ``Success: no issues found in 1 source file``

**Actual Behavior**

I get this error:
```
main.py:21: error: Item "None" of "Path | None" has no attribute "exists"  [union-attr]
Found 1 error in 1 file (checked 1 source file)
```

Note that if I comment the line ``b = a``, I don't get an error.
I believe it is an error because since there is already a check to verify that ``a.path`` isn't ``None`` before ``b = a``, so it implies that ``b.path`` cannot be ``None``.

**Your Environment**

- Mypy version used: 1.19.1
- Mypy command-line flags: ``mypy main.py``
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.13.9

PS: Sorry if this is a duplicate issue. I didn't really know how to "name" this problem, so I didn't find if there was already a bug report or not.

**Comment by hauntsaninja** [comment]
Mypy doesn't track relationships across variables like this. All other Python type checkers I am aware of behave similarly. Preserving narrowing across aliasing also opens up space for more unsoundness in the case of mutation.

There are a few duplicates of this in the issue tracker, but it is indeed a tricky thing to "name" :-)
Closing as a "won't fix" / "as currently designed"
